include/drivers/rtio.h

```diff
@@ -0,0 +1,487 @@
+/*
+ * Copyright (c) 2019  Thomas Burdick <thomas.burdick@gmail.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_INCLUDE_DRIVERS_RTIO_H_
+#define ZEPHYR_INCLUDE_DRIVERS_RTIO_H_
+/**
+ * @brief RTIO Interface
+ * @defgroup rtio_interface RTIO Interface
+ * @ingroup io_interfaces
+ * @{
+ */
+
+#include <kernel.h>
+#include <device.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @file
+ * @brief Real-Time IO device API for moving bytes around fast
+ *
+ * Each Real-Time IO device provides a way to either read, write, or both
+ * blocks of memory allocated from a static pool by the application. The layout
+ * of the each block is understood implicitly by the author. Each device may
+ * provide functionality to safely encode or decode blocks of memory.
+ */
+
+/**
+ * @brief Block of memory
+ *
+ * This block of memory has important metadata commonly useful such as
+ * byte layout, and timestamps of when a read or write began and ended.
+ */
+struct rtio_block {
+	/** Timestamp in cycles from k_cycle_get_32() marking when a read or
+	 * write began
+	 */
+	u32_t begin_tstamp;
+
+	/** Timestamp in cycles from k_cycle_get_32() marking when a read or
+	 * write was complete
+	 */
+	u32_t end_tstamp;
+
+	/** Byte layout designator given by each driver to specify
+	 * the layout of the contained buffer
+	 */
+	u32_t layout;
+
+	/** Length of data in this buffer. */
+	u16_t len;
+
+	/** Amount of data this buffer can store. */
+	u16_t size;
+
+	/** Start of the buffer data .*/
+	u8_t *_data;
+};
+
+/**
+ * @brief Initialize an rtio_block
+ */
+static inline void rtio_block_init(struct rtio_block *block, u8_t *data,
+				   u32_t size)
+{
+	block->layout = 0;
+	block->begin_tstamp = 0;
+	block->end_tstamp = 0;
+	block->len = 0;
+	block->size = size;
+	block->_data = data;
+}
+
+/**
+ * @brief Begin a block read or write
+ */
+static inline void rtio_block_begin(struct rtio_block *block)
```

Comment:
Excellent, yes, and in fact it would be fantastic if I can get the precision of the clock in the future as in some instances I'd like to interpolate timestamps of sample sets as a future PR will show

Reply:
As the clock source is not yet in, I believe this should be ok until that happens